### PR TITLE
Convert a grab bag of database code to async/await

### PR DESCRIPTION
**Pull request by clokep**
Spun out of #8156, this converts a variety of the database classes to async/await: transactions, appservice, group_server, deviceinbox, e2e_room_keys, keys, event_federation.

Not really any particular relation between these, but I limited myself on size.

### CODE REVIEW

File: synapse/appservice/__init__.py [hunk 1]
```diff
@@ -35,30 +35,28 @@ def __init__(self, service, id, events):
         self.id = id
         self.events = events
 
-    def send(self, as_api):
+    async def send(self, as_api) -> bool:
```
Comment (erikjohnston) [inline synapse/appservice/__init__.py:38]: I'm slightly in two minds about filling out partial type hints in a function, in terms of making it less obvious which functions are typed or not. I could go either way though

File: synapse/storage/databases/main/transactions.py [hunk 2]
```diff
@@ -66,11 +66,11 @@ def get_received_txn_response(self, transaction_id, origin):
             origin(str)
 
         Returns:
-            tuple: None if we have not previously responded to
+            Tuple|None: None if we have not previously responded to
```
Comment (erikjohnston) [inline synapse/storage/databases/main/transactions.py:69]: We may as well move this up into type hints if we're correcting things

**Review by erikjohnston — APPROVED** [review]

File: synapse/federation/persistence.py [hunk 3]
```diff
@@ -36,19 +36,21 @@ def __init__(self, datastore):
...
 
     @log_function
-    def have_responded(self, origin, transaction):
+    async def have_responded(self, origin, transaction):
         """ Have we already responded to a transaction with the same id and
         origin?
 
         Returns:
-            Deferred: Results in `None` if we have not previously responded to
+            Awaitable: Results in `None` if we have not previously responded to
```
[the remark this answers is not in the document]
Comment (clokep) [reply synapse/federation/persistence.py:44]: Oh yes, I must have missed this one in my rebase. 👍

**Comment by clokep** [comment]
I went back through the modified methods and added type information where it was readily available (and ensured there wasn't partial information).

**Pull request merged**